src/Processors/QueryPlan/ReadFromRemote.cpp

```diff
@@ -324,42 +320,49 @@ void ReadFromParallelRemoteReplicasStep::initializePipeline(QueryPipelineBuilder
     auto timeouts = ConnectionTimeouts::getTCPTimeoutsWithFailover(current_settings);
 
     size_t all_replicas_count = current_settings.max_parallel_replicas;
-    if (all_replicas_count > shard_info.all_addresses.size())
+    if (all_replicas_count > cluster->getShardsInfo().size())
     {
         LOG_INFO(&Poco::Logger::get("ReadFromParallelRemoteReplicasStep"),
             "The number of replicas requested ({}) is bigger than the real number available in the cluster ({}). "\
-            "Will use the latter number to execute the query.", current_settings.max_parallel_replicas, shard_info.all_addresses.size());
-        all_replicas_count = shard_info.all_addresses.size();
+            "Will use the latter number to execute the query.", current_settings.max_parallel_replicas, cluster->getShardsInfo().size());
+        all_replicas_count = cluster->getShardsInfo().size();
+    }
+
+    /// Find local shard
+    for (const auto & shard: cluster->getShardsInfo())
+    {
+        if (shard.isLocal())
+        {
+            IConnections::ReplicaInfo replica_info
+            {
+                .all_replicas_count = all_replicas_count,
+                .number_of_current_replica = 0
+            };
+
+            addPipeForSingeReplica(pipes, shard.pool, replica_info);
+        }
     }
 
-    /// The requested number of replicas to read from could be less
-    /// than the total number of replicas in the shard
-    /// And we have to pick only "remote" ones
-    /// So, that's why this loop looks like this.
-    size_t replica_num = 0;
-    while (pipes.size() != all_replicas_count - 1)
+    if (pipes.empty())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "No local shard");
```

Comment:
`LOGICAL_EROR` doesn't look right. It is possible for a cluster doesn't have a replica tagged as is_local (because default database is declared) for example

cc @nikitamikhaylov

Reply:
Related ticket: https://github.com/ClickHouse/ClickHouse/issues/46277 Although this has been the CH behaviour for a really long time

Reply:
Hmm, Ok. In general it doesn't matter, so we can remove this check and just handle the absence of local replica.